commit 8efc8de8d0fc3c617d955c5d1a59b5f321b7511f
Author: Zijie Yan <zijiey@nvidia.com>
Date:   Tue Dec 26 11:46:32 2023 +0000

    Fix moe aux loss.

diff --git a/megatron/core/transformer/moe/base_moe_layer.py b/megatron/core/transformer/moe/base_moe_layer.py
index 5e18c0e10..c5d9ca6a8 100644
--- a/megatron/core/transformer/moe/base_moe_layer.py
+++ b/megatron/core/transformer/moe/base_moe_layer.py
@@ -117,11 +117,11 @@ class Router(ABC, MegatronModule):
 
         return scores, indices
 
-    def apply_aux_loss(self, loss_func, scores, indicies):
-        mask = torch.nn.functional.one_hot(indicies, num_classes=self.num_experts).sum(dim=1)
-        aux_loss = loss_func(scores, mask)
-        scores = MoEAuxLossAutoScaler.apply(scores, aux_loss)
-        return scores
+    def apply_aux_loss(self, loss_func, probs, indices):
+        mask = torch.nn.functional.one_hot(indices, num_classes=self.num_experts).sum(dim=1)
+        aux_loss = loss_func(probs, mask, self.config.moe_aux_loss_coeff)
+        indices = MoEAuxLossAutoScaler.apply(indices, aux_loss)
+        return indices
 
     def apply_z_loss(self, logits):
         """Encourages the router's logits to remain small to enhance stability.
@@ -500,19 +500,19 @@ class ZeroDropTopKRouter(Router):
         """
         logits = logits.view(-1, self.config.num_moe_experts)
         logits = logits.to(dtype=torch.float32)
-        logits = torch.softmax(logits, dim=-1)
+        probs = torch.softmax(logits, dim=-1)
 
         # Apply Z-Loss
         if self.config.moe_z_loss_coeff > 0:
-            logits = self.apply_z_loss(logits)
+            probs = self.apply_z_loss(probs)
 
-        scores, indices = torch.topk(logits, k=self.k, dim=1)
+        scores, indices = torch.topk(probs, k=self.k, dim=1)
 
         scores /= scores.sum(dim=-1, keepdim=True)
 
         # Apply load balancing loss
         if self.config.moe_aux_loss_coeff > 0:
-            scores = self.apply_aux_loss(self.moe_aux_loss_func, scores, indices)
+            indices = self.apply_aux_loss(self.moe_aux_loss_func, probs, indices)
 
         return scores, indices
 
diff --git a/megatron/core/transformer/moe/moe_utils.py b/megatron/core/transformer/moe/moe_utils.py
index 04a53d021..938324933 100644
--- a/megatron/core/transformer/moe/moe_utils.py
+++ b/megatron/core/transformer/moe/moe_utils.py
@@ -1,7 +1,7 @@
 import torch
 
 
-def switch_load_balancing_loss_func(config, gates, mask):
+def switch_load_balancing_loss_func(gates, mask, moe_aux_loss_coeff):
     """Calculate the auxiliary loss for better load balacing. 
     Please refer to the Switch Transformer paper (https://arxiv.org/abs/2101.03961) for details.
 
@@ -12,12 +12,11 @@ def switch_load_balancing_loss_func(config, gates, mask):
     Returns:
         torch.Tensor: The auxiliary loss for load balancing.
     """
-    num_experts = mask.size(1)
-    assert num_experts == config.num_moe_experts
+    num_experts = mask.size(-1)
     gates_mean = gates.mean(dim=0)
     selection_mean = mask.float().mean(dim=0)
     aux_loss = torch.sum(gates_mean * selection_mean) * num_experts
-    aux_loss *= config.aux_loss_coeff
+    aux_loss *= moe_aux_loss_coeff
     return aux_loss
 
 
