From 5486c69c627e98530dbc556e5c404fed2258b311 Mon Sep 17 00:00:00 2001
From: Mikołaj Błaż <mblaz@nvidia.com>
Date: Tue Apr 9 11:23:39 2024 +0200
Subject: Add debug times

diff --git a/megatron/core/dist_checkpointing/strategies/fully_parallel.py b/megatron/core/dist_checkpointing/strategies/fully_parallel.py
index d3a672a18..1cd9231cf 100644
--- a/megatron/core/dist_checkpointing/strategies/fully_parallel.py
+++ b/megatron/core/dist_checkpointing/strategies/fully_parallel.py
@@ -132,7 +132,11 @@ class FullyParallelLoadStrategyWrapper(LoadShardedStrategy):
         if torch.distributed.get_world_size(self.parallelization_group) <= 1:
             return self.base_strategy.load(sharded_state_dict, checkpoint_dir)
 
+        start = time()
         precomputed_distribution = self.apply_loading_parallelization(sharded_state_dict)
+        end = time()
+        logger.debug(f'self.apply_loading_parallelization took {end - start}s')
+        start = end
         (
             sharded_tensors,
             sharded_state_dict,
@@ -142,9 +146,18 @@ class FullyParallelLoadStrategyWrapper(LoadShardedStrategy):
         # Load only sharded objects
         loaded_state_dict = self.base_strategy.load(sharded_state_dict, checkpoint_dir)
 
+        end = time()
+        logger.debug(f'Base load of ShardedObjects took {end - start}s')
+        start = end
+
         # Load sharded tensors separately
-        print(f'Applying parallel load with algo {self.gather_algo}')
         loaded_tensors = self.base_strategy.load(to_load_shards, checkpoint_dir)
+
+        end = time()
+        logger.debug(f'Base load of ShardedTensors took {end - start}s')
+        start = end
+
+        logger.debug(f'Applying parallel load with algo {self.gather_algo}')
         if self.gather_algo == 'object':
             all_loaded_tensors = self.exchange_loaded_tensors_gather_object(
                 loaded_tensors, unloaded_shards, precomputed_distribution, self.parallelization_group
@@ -155,6 +168,13 @@ class FullyParallelLoadStrategyWrapper(LoadShardedStrategy):
             )
         else:
             raise NotImplementedError(f'Unrecognized gather algorithm: {self.gather_algo}')
+
+        sync_start = time()
+        torch.cuda.synchronize()
+        end = time()
+        logger.debug(f'torch.cuda.synchronize took {end - sync_start}s')
+        logger.debug(f'self.exchange_loaded_tensors took {end - start}s')
+
         self.fill_in_deferred_sharded_tensors(sharded_tensors, all_loaded_tensors)
         merge(loaded_state_dict, sharded_tensors)
         return loaded_state_dict
